crates/sui-json-rpc/src/authority_state.rs

```diff
@@ -413,15 +474,99 @@ impl StateRead for AuthorityState {
         limit: usize,
         one_coin_type_only: bool,
     ) -> StateReadResult<Vec<SuiCoin>> {
-        Ok(self
-            .get_owned_coins_iterator_with_cursor(owner, cursor, limit, one_coin_type_only)?
+        // Get the currency types (e.g. `0x2::sui::SUI`) for all relevant address balances.
+        let address_balance_currency_types = if !one_coin_type_only {
+            let child_object_resolver = self.get_child_object_resolver().as_ref();
+
+            get_currency_types_for_owner(
+                owner,
+                child_object_resolver,
+                self.indexes
+                    .as_ref()
+                    .ok_or(SuiErrorKind::IndexStoreNotAvailable)?
+                    .tables(),
+            )?
+            .into_iter()
+            // ignore any types that are less than the type in the cursor
+            .filter(|t| t.to_string() >= cursor.0)
+            .collect()
+        } else {
+            // This error should be unreachable (the type has already been parsed earlier).
+            vec![
+                parse_to_type_tag(Some(dbg!(cursor.0.clone())))
+                    .map_err(|_| anyhow::anyhow!("Invalid coin type: {}", cursor.0))?,
+            ]
+        };
+
+        dbg!(&address_balance_currency_types);
+
+        let cursor_key =
+            CoinIndexKey2::new_from_cursor(owner, cursor.0.clone(), cursor.1, cursor.2);
+
+        // load all active address balances for the address. While we repeat this work on every call
+        // for a paginated query, the number of distinct address balances will typically be small, and
+        // redundant reads will be served from cache
+        let mut address_balance_coins = vec![];
+        for currency_type in address_balance_currency_types {
+            if let Some((obj_ref, balance, previous_transaction)) =
+                self.get_address_balance_coin_info(owner, Balance::type_tag(currency_type.clone()))?
+            {
+                let key = CoinIndexKey2::new(owner, currency_type.to_string(), balance, obj_ref.0);
+                if key < cursor_key {
+                    continue;
+                }
+                address_balance_coins.push((
+                    key,
+                    CoinInfo {
+                        version: obj_ref.1,
+                        digest: obj_ref.2,
+                        balance,
+                        previous_transaction,
+                    },
+                ));
+            }
+        }
+
+        dbg!(&address_balance_coins);
+
+        address_balance_coins.sort();
+
+        let mut coin_iter = self
+            .get_owned_coins_iterator_with_cursor(owner, cursor.clone(), limit, one_coin_type_only)?
+            .peekable();
+        let mut address_balance_coins_iter = dbg!(address_balance_coins).into_iter().peekable();
+        let mut merged_coins = vec![];
+        loop {
+            match (address_balance_coins_iter.peek(), coin_iter.peek()) {
```

Comment:
This merge sort works only when `address_balance_coins_iter` and `coin_iter` are both already sorted by their keys. `coin_iter` is sorted because it comes directly from the `coin_index_2` index with that same key. But is `address_balance_coins_iter` sorted by `CoinIndexKey2`? The order of `address_balance_coins_iter` comes from `get_currency_types_for_owner` where we fetch from the dynamic field index, which is keyed by `(parent_id, child_id)`.